### PR TITLE
[WIP] Unify sampling

### PR DESCRIPTION
**Pull request by josh146**
**Context:** Currently, a single QNode evaluation corresponds to a single 'device' execution. However, hardware devices will always return correlated samples per evaluation, whereas simulators return uncorrelated samples.

**Description of the Change:** Modifies `QubitDevice` and `DefaultQubit` so that multiple sample calls from simulators within a single evaluation return correlated results.

**Benefits:** Consistency.

**Possible Drawbacks:** n/a

**Related GitHub Issues:** n/a

**Todo:**

- [x] Double-check and rethink the logic flow
- [x] Write new unit tests for the new methods in `QubitDevice`
- [X] Write tests to ensure that correlated samples happen in `DefaultQubit`
- [x] Re-write the sample integration tests (currently marked xfail)
- [x] Reorganize tests of `DefaultQubit` into `QubitDevice`'s test file

### CODE REVIEW

File: tests/test_tensor_measurements.py [hunk 1]
```diff
@@ -318,6 +318,7 @@ def circuit(a, b, c):
         assert np.allclose(res, expected, atol=tol, rtol=0)
 
 
+@pytest.mark.xfail
```
Comment (josh146) [inline tests/test_tensor_measurements.py]: These tests are currently failing and will need to be rewritten. As they are integration tests, I am okay with them being stochastic (i.e., look at the resulting samples, take the mean and variance, and make sure they match <A> and var(A)). However, care should be taken so that the chance they fail stochastically is so small, it will likely never happen.

File: pennylane/_qubit_device.py [hunk 2]
```diff
@@ -38,11 +38,15 @@ class QubitDevice(Device):
             expectation values of observables. Defaults to 1000 if not specified.
     """
     #pylint: disable=too-many-public-methods
-    def __init__(self, wires=1, shots=1000, analytic=True):
+    def __init__(self, wires=1, shots=1000, analytic=True, rotate_basis=True):
```
Comment (antalszava) [inline pennylane/_qubit_device.py]: As per our discussion, ``rotate_basis`` was added such that devices that natively support expectation values do not rotate the circuit.

However, currently, I would argue that from the set of devices that are supported, neither are such devices. As this change does not seem to be too complicated, it might be something that would be worth adding, once there was a need indeed for this functionality.

File: pennylane/_qubit_device.py [hunk 3]
```diff
@@ -119,44 +126,96 @@ def execute(self, queue, observables, parameters=None):
 
             return self._asarray(results)
 
-    def rotate_basis(self, observable):
+    def rotate_basis(self, obs_queue):
         """Rotates the specified wires such that they
         are in the eigenbasis of the provided observable.
+        """
+        self._prob = self.probability(values_only=True)
```
Comment (antalszava) [inline pennylane/_qubit_device.py]: We had a discussion about this attribute, could you remind me what the devices could use this attribute for?

Comment (josh146) [reply pennylane/_qubit_device.py]: This is a bit hacky, and it's only used to avoid creating that ordered dictionary (which would increase memory usage for very large systems)

Comment (josh146) [reply pennylane/_qubit_device.py]: Ideally, we would implement a better return type for `probability`, such as a custom class.

File: pennylane/_qubit_device.py (hunk 2 above)
Comment (josh146) [reply pennylane/_qubit_device.py]: Yep, sounds good :+1:

File: pennylane/plugins/default_qubit.py [hunk 4]
```diff
@@ -380,6 +379,10 @@ def apply(self, operation):
         self._state = self.mat_vec_product(A, self._state, wires)
         self._first_operation = False
 
+    def post_apply(self):
+        # generate computational basis samples
+        self.generate_samples()
+
```
Comment (antalszava) [inline pennylane/plugins/default_qubit.py]: How come this was only implemented for ``default.qubit``?

Also, would there be any reason not to move ``rotate_basis`` into ``post_apply``?

One of the things HW devices could do is to override the ``generate_samples`` method used in ``post_apply``. One of the advantages of that would be, that actually the main difference between a HW device and a statevector simulator would be is how they generate samples.

Comment (antalszava) [reply pennylane/plugins/default_qubit.py]: Would it also perhaps be better to move these into ``pre_measure``? Currently (as there was discussion about this) it might still be ambiguous what is happening in each of the ``pre``/``post`` functions

Comment (josh146) [reply pennylane/plugins/default_qubit.py]: These are good points. I'm starting to think `generate_samples` shouldn't be a surprise to the plugin developer, which it might be if it is within `pre_measure`/`post_apply`. Maybe the plugin needs to explicitly call `generate_samples`, or explicitly provide a `bool` that turns this feature on and off.

Since `post_apply`, `pre_measure` etc. don't really have any definition (there convenience methods for the plugin to overwrite if needed), maybe it's a bad idea to put things there by default?

File: pennylane/_qubit_device.py (hunk 3 above)
Comment (antalszava) [reply pennylane/_qubit_device.py]: Provided that the ``basis_states`` are always created based on the order coming from
```
        basis_states = itertools.product(range(2), repeat=len(wires))
```
(which could be documented), we could actually always just return ``probability``, correct?

The user can also just construct ``basis_states`` or we could even have a ``property`` for that such as
```python
@property
def probabilitiy_with_basis_states(self):
    basis_states = itertools.product(range(2), repeat=len(wires))
    return dict(zip(basis_states, self.probability))
```
This can be done like this, as [``itertools.product``](https://docs.python.org/3.8/library/itertools.html#itertools.product) keeps ordering:
> Roughly equivalent to nested for-loops in a generator expression. For example, product(A, B) returns the same as ((x,y) for x in A for y in B).

File: tests/test_qubit_device.py [hunk 5]
```diff
@@ -0,0 +1,1121 @@
...
+            num_wires = 1
+            par_domain = 'F'
+            return_type = returntype
+
+        obs = SomeObservable(wires=0)
+
+        with pytest.raises(
+                QuantumFunctionError, match="Unsupported return type"
+        ):
+           results = mock_qubit_device_extract_stats.statistics([obs])
```
Comment (josh146) [inline tests/test_qubit_device.py]: Nice!

File: tests/test_qubit_device.py [hunk 6]
```diff
@@ -0,0 +1,1121 @@
...
+                             )
+                               ),
+                            (
+                            np.array([2, 3, 1, 3, 1]),
+                            np.array([[0, 1, 0, 0],
+                                       [1, 1, 0, 0],
+                                       [1, 0, 0, 0],
+                                       [1, 1, 0, 0],
+                                       [1, 0, 0, 0]])
+
```
Comment (josh146) [inline tests/test_qubit_device.py]: Maybe I am just getting a bit confused, but don't we use the convention in PennyLane that computational basis state 2 corresponds to |0010>, not |0100>?

File: pennylane/_qubit_device.py [hunk 7]
```diff
@@ -112,83 +112,237 @@ def execute(self, queue, observables, parameters=None):
...
+    def rotate_basis(self, obs_queue):
         """Rotates the specified wires such that they
         are in the eigenbasis of the provided observable.
 
         Args:
-            observable (Observable): the observable that is to be
-                measured
+            observables (List[:class:`Observable`]): the observable queue
+        """
+        self._prob = self.probability()
```
Comment (josh146) [inline pennylane/_qubit_device.py]: This doesn't seem to be used anywhere :thinking:

File: pennylane/_qubit_device.py [hunk 8]
```diff
@@ -112,83 +112,237 @@ def execute(self, queue, observables, parameters=None):
...
+    def sample(self, observable):
+        wires = observable.wires
+        name = observable.name
+
+        if self._rotated_prob is None:
+            observable.return_type = Sample
+            self.rotate_basis([observable])
+
+            if self._memory or (not self.analytic):
+                self.generate_samples()
```
[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane/_qubit_device.py]: Added TODOs to each place and opened the #468 issue!

File: tests/test_qubit_device.py [hunk 9]
```diff
@@ -0,0 +1,1121 @@
...
+        m.setattr(QubitDevice, 'short_name', 'MockDevice')
+        m.setattr(QubitDevice, 'expval', lambda self, x: 0)
+        m.setattr(QubitDevice, 'var', lambda self, x: 0)
+        m.setattr(QubitDevice, 'sample', lambda self, x: 0)
+        m.setattr(QubitDevice, 'apply', lambda self, x: None)
+        yield QubitDevice()
+
+@pytest.fixture(scope="function")
+def mock_qubit_device_extract_stats(monkeypatch):
+    with monkeypatch.context() as m:
```
[the remark this answers is not in the document]
Comment (antalszava) [reply tests/test_qubit_device.py]: Good idea!

File: tests/test_tensor_measurements.py [hunk 10]
```diff
@@ -317,41 +320,62 @@ def circuit(a, b, c):
...
-            + 2 * np.cos(2 * theta)
-            + 2 * np.cos(2 * phi)
-            + 14
-        ) / 16
-        assert np.allclose(var, expected, atol=tol, rtol=0)
+        # Diagonalize according to the observable
+        psi = tensor_product([H, I, I]) @ psi
+        psi = tensor_product([I, I, Z]) @ psi
+        psi = tensor_product([I, I, S]) @ psi
+        psi = tensor_product([I, I, H]) @ psi
```
[the remark this answers is not in the document]
Comment (antalszava) [reply tests/test_tensor_measurements.py]: Yes, that would actually be awesome!

File: tests/test_tensor_measurements.py [hunk 11]
```diff
@@ -317,41 +320,62 @@ def circuit(a, b, c):
...
+        """Test that a tensor product involving PauliX and PauliZ works correctly"""
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.Hadamard(wires=0)
+            return sample(qml.PauliX(0) @ qml.PauliZ(1))
+
+        with monkeypatch.context() as m:
+            s1 = circuit()
```
[the remark this answers is not in the document]
Comment (antalszava) [reply tests/test_tensor_measurements.py]: Corrected each of them, thanks so much for that!

File: tests/test_tensor_measurements.py [hunk 12]
```diff
@@ -317,41 +320,62 @@ def circuit(a, b, c):
...
-        ) / 16
-        assert np.allclose(var, expected, atol=tol, rtol=0)
+        # Diagonalize according to the observable
+        psi = tensor_product([H, I, I]) @ psi
+        psi = tensor_product([I, I, Z]) @ psi
+        psi = tensor_product([I, I, S]) @ psi
+        psi = tensor_product([I, I, H]) @ psi
+
+        expected_probabilities = np.abs(psi) ** 2
+        expected_probabilities = expected_probabilities.tolist()
```
[the remark this answers is not in the document]
Comment (antalszava) [reply tests/test_tensor_measurements.py]: Oh, it remained from another state of the test, removing it! Thanks :)

File: tests/test_qubit_device.py (hunk 6 above)
Comment (antalszava) [reply tests/test_qubit_device.py]: (As per our discussion)
It was totally confusing for me as well now that I saw it! :laughing: But so this one is used the following way in the code:
```python
self._samples[:, wires[0]]
```
Meaning that contrary to the mathematical representation, the first qubit will be the first columns, second qubit the second column, etc.

Thanks for the note, I'll add a comment regarding this!

File: pennylane/_qubit_device.py (hunk 7 above)
Comment (antalszava) [reply pennylane/_qubit_device.py]: Correct! This one was added such that the probability before the rotation is stored and can be potentially retrieved

Comment (johannesjmeyer) [reply pennylane/_qubit_device.py]: What was the use case you had in mind for this pre-rotation probability?

File: pennylane/_qubit_device.py [hunk 13]
```diff
@@ -112,83 +112,246 @@ def execute(self, queue, observables, parameters=None):
...
+                self.apply(diag_gate)
+
+            for wire in observable.wires:
+                if isinstance(wire, int):
+                    wires.append(wire)
+                else:
+                    wires.extend(wire)
+
+        self._wires_used = wires
+        self._rotated_prob = self.probability(wires)
```
Comment (johannesjmeyer) [inline pennylane/_qubit_device.py]: One could think about not calculating the probabilities if there are only samples requested from the circuit

File: pennylane/_qubit_device.py [hunk 14]
```diff
@@ -112,83 +112,246 @@ def execute(self, queue, observables, parameters=None):
...
+            for diag_gate in observable.diagonalizing_gates():
+                self.apply(diag_gate)
+
+            for wire in observable.wires:
+                if isinstance(wire, int):
+                    wires.append(wire)
+                else:
+                    wires.extend(wire)
+
+        self._wires_used = wires
```
Comment (johannesjmeyer) [inline pennylane/_qubit_device.py]: What is the point of `_wires_used`?

As `rotate_basis` is also called inside `expval` for example, this value will be reset continually. If other methods then depend on it this might cause unexpected behaviour.

File: pennylane/_qubit_device.py [hunk 15]
```diff
@@ -112,83 +112,246 @@ def execute(self, queue, observables, parameters=None):
...
+        # unit tests directly calling on sample()
+        if self._rotated_prob is None:
+            observable.return_type = Sample
+            self.rotate_basis([observable])
+
+            if self._memory or (not self.analytic):
+                self.generate_samples()
 
-    def probability(self, wires=None, values_only=False):
+        if isinstance(name, str) and name in {"PauliX", "PauliY", "PauliZ", "Hadamard"}:
```
Comment (johannesjmeyer) [inline pennylane/_qubit_device.py]: Is there such a dramatic increase in performance by having the Pauli expvals in a separate (more efficient?) method?

Comment (johannesjmeyer) [reply pennylane/_qubit_device.py]: Okay, when looking at the code, I guess there is!

File: pennylane/plugins/default_qubit.py (hunk 4 above)
Comment (johannesjmeyer) [reply pennylane/plugins/default_qubit.py]: I also think that the `default.qubit` plugin should overwrite `generate_samples` so that samples are only generated if it is not in analytic mode.

Comment (antalszava) [reply pennylane/plugins/default_qubit.py]: That's a fair point! In the cases when ``analytic`` is ``False`` and the ``return_type`` is not ``Sample``, ``default.qubit`` shouldn't generate samples

File: pennylane/_qubit_device.py (hunk 7 above)
Comment (antalszava) [reply pennylane/_qubit_device.py]: The notion behind this was to be able to perform sanity checks by looking at the circuits. Ultimately it might be better to store the state in fashion as ``dev._state`` does

File: pennylane/_qubit_device.py (hunk 13 above)
Comment (antalszava) [reply pennylane/_qubit_device.py]: As the previous attribute, this would also mostly be something that could be used for sanity check. One samples are being returned, this would be equal to having the probabilities before computational basis samples are being extracted

File: pennylane/_qubit_device.py (hunk 7 above)
Comment (josh146) [reply pennylane/_qubit_device.py]: Yep, I agree. Probability can now be returned via the QNode, so it makes more sense to cache the pre-rotation state.

File: pennylane/_qubit_device.py (hunk 14 above)
Comment (josh146) [reply pennylane/_qubit_device.py]: > What is the point of _wires_used?

I'll let @antalszava answer this one!

> As rotate_basis is also called inside expval for example

Note that calling `rotate_basis` inside `expval` is a quick 'hack' to make the existing default.qubit test suite pass (as it doesn't have a concept of rotation, and simply calls `dev.expval`). The plan is to remove this call inside `expval`, and refactor the tests to take this into account.

Comment (antalszava) [reply pennylane/_qubit_device.py]: The point of ``_wires_used`` is to keep track of the wires where an observable is applied. Those wires, that are not in ``_wires_used`` (if any) will have the ``Identity`` applied implicitly and probabilities will need to be summed over such wires once marginal probabilities are returned.

It is good feedback to see that it is not entirely clear how this is used, will try to rename it.

Currently   ``rotate_basis`` is only called inside of e.g. ``expval`` if the following statement evaluates as ``True``:
```python
if self._rotated_prob is None:
```
This logic will be refactored at a later stage, currently it is only included such that tests can be passed easily (it would require a major test refactor).

Comment (johannesjmeyer) [reply pennylane/_qubit_device.py]: Okay, I see!

Comment (johannesjmeyer) [reply pennylane/_qubit_device.py]: And from my point of view, `_wires_used` is a pretty good name, no need to change that. I was just wondering if the `_wires_used` attribute is still sane if it is modified in expval

File: pennylane/_qubit_device.py (hunk 7 above)
Comment (johannesjmeyer) [reply pennylane/_qubit_device.py]: Okay, I see. But is this something intended for internal use only or is there a special UI for accessing the probabilities pre-rotation?

Comment (josh146) [reply pennylane/_qubit_device.py]: I think this is something that it outdated, and will be changed to instead cache the *state* pre rotation.

File: pennylane/_qubit_device.py [hunk 16]
```diff
@@ -178,12 +175,17 @@ def rotate_basis(self, obs_queue):
...
+        # Store the wires used by the observables such that
+        # an Identity is considered on the remaining wires
         self._wires_used = wires
         self._rotated_prob = self.probability(wires)
 
     def generate_samples(self):
         """Generate computational basis samples based on the current state.
 
+        Warning: this method will have to be redefined for hardware devices, since it uses
+        the device._rotated_prob. This attribute might not be available for such devices.
```
Comment (josh146) [inline pennylane/_qubit_device.py]: Note: you can use `.. warning::` here to make a nice yellow box. Also, I would rephrase this - `device._rotated_prob` is set internally by the `QubitDevice` by simply calling `self.probability`. The reason this should be redefined for hardware devices: because hardware devices generate samples themselves :)

File: pennylane/_qubit_device.py [hunk 17]
```diff
@@ -323,21 +326,19 @@ def probability(self, wires=None):
         """Return the (marginal) probability of each computational basis
         state from the last run of the device.
 
+        Warning: this method will have to be redefined for hardware devices, since it uses
+        the device._state attribut. This attribute might not be available for such devices.
```
Comment (josh146) [inline pennylane/_qubit_device.py]: `attribute`

File: pennylane/_qubit_device.py (hunk 16 above)
Comment (antalszava) [reply pennylane/_qubit_device.py]: Ah you are right, thanks! Also, I just saw that there is already a warning included down below, so I'll just leave that as included (instead of having a new one defined :smile:).
